property/collector_test.go

```diff
@@ -0,0 +1,316 @@
+/*
+Copyright (c) 2024-2024 VMware, Inc. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package property_test
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/vmware/govmomi/find"
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/property"
+	"github.com/vmware/govmomi/simulator"
+	"github.com/vmware/govmomi/vim25"
+	"github.com/vmware/govmomi/vim25/types"
+)
+
+func TestWaitForUpdatesEx(t *testing.T) {
+	model := simulator.VPX()
+	model.Datacenter = 1
+	model.Cluster = 0
+	model.Pool = 0
+	model.Machine = 1
+	model.Autostart = false
+
+	simulator.Test(func(ctx context.Context, c *vim25.Client) {
+		// Set up the finder and get a VM.
+		finder := find.NewFinder(c, true)
+		datacenter, err := finder.DefaultDatacenter(ctx)
+		if err != nil {
+			t.Fatalf("default datacenter not found: %s", err)
+		}
+		finder.SetDatacenter(datacenter)
+		vmList, err := finder.VirtualMachineList(ctx, "*")
+		if len(vmList) == 0 {
+			t.Fatal("vmList == 0")
+		}
+		vm := vmList[0]
+
+		pc, err := property.DefaultCollector(c).Create(ctx)
+		if err != nil {
+			t.Fatalf("failed to create new property collector: %s", err)
+		}
+
+		// Start a goroutine to wait for power state changes to the VM.
+		chanResult := make(chan any)
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		go func() {
+			defer close(chanResult)
+			if err := property.WaitForUpdatesEx(
+				ctx,
+				pc,
+				&property.WaitFilter{
+					CreateFilter: getDatacenterToVMFolderFilter(datacenter),
+					WaitOptions: property.WaitOptions{
+						Options: &types.WaitOptions{
+							MaxWaitSeconds: addrOf(int32(3)),
+						},
+					},
+				},
+				func(updates []types.ObjectUpdate) bool {
+					return waitForPowerStateChanges(
+						ctx,
+						vm,
+						chanResult,
+						updates,
+						types.VirtualMachinePowerStatePoweredOn)
+				},
+			); err != nil {
+				chanResult <- err
+				return
+			}
+		}()
+
+		// Power on the VM to cause a property change.
+		if _, err := vm.PowerOn(ctx); err != nil {
+			t.Fatalf("error while powering on vm: %s", err)
+		}
+
+		select {
+		case <-time.After(3 * time.Second):
+			t.Fatalf("timed out while waiting for property update")
+		case result := <-chanResult:
+			switch tResult := result.(type) {
+			case types.VirtualMachinePowerState:
+				if tResult != types.VirtualMachinePowerStatePoweredOn {
```

Comment:
@dougm I was stuck on this for several hours yesterday. Well, an earlier version of it. It was when I realized that the value of the change was _not_ a string, but a `VirtualMachinePowerState` :)